internal/coord/routing/crawl.go

```diff
@@ -0,0 +1,357 @@
+// Options may be overridden before passing to NewCrawl
+func DefaultCrawlConfig() *CrawlConfig {
+	return &CrawlConfig{
+		MaxCPL:      16,
+		Concurrency: 200,
+		Tracer:      tele.NoopTracer(),
+	}
+}
+
+type Crawl[K kad.Key[K], N kad.NodeID[K]] struct {
+	self  N
+	cfg   CrawlConfig
+	cplFn coordt.NodeIDForCplFunc[K, N]
+	info  *crawlInformation[K, N] // only set of crawl is in progress
+}
+
+type crawlInformation[K kad.Key[K], N kad.NodeID[K]] struct {
+	todo    []crawlJob[K, N]
+	cpls    map[string]int
+	waiting map[string]N
+	success map[string]N
+	failed  map[string]N
+	errors  map[string]error
+}
+
+func NewCrawl[K kad.Key[K], N kad.NodeID[K]](self N, cplFn coordt.NodeIDForCplFunc[K, N], cfg *CrawlConfig) (*Crawl[K, N], error) {
+	if cfg == nil {
+		cfg = DefaultCrawlConfig()
+	} else if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	c := &Crawl[K, N]{
+		self:  self,
+		cfg:   *cfg,
+		cplFn: cplFn,
+		info:  nil,
+	}
+
+	return c, nil
+}
+
+func (c *Crawl[K, N]) Advance(ctx context.Context, ev CrawlEvent) (out CrawlState) {
+	_, span := c.cfg.Tracer.Start(ctx, "Crawl.Advance", trace.WithAttributes(tele.AttrInEvent(ev)))
+	c.setMapSizes(span, "before")
+	defer func() {
+		c.setMapSizes(span, "after")
+		span.SetAttributes(tele.AttrOutEvent(out))
+		span.End()
+	}()
+
+	switch tev := ev.(type) {
+	case *EventCrawlStart[K, N]:
+		if c.info != nil {
+			break // query in progress, pretend it was a poll
+		}
+
+		span.SetAttributes(attribute.Int("seed", len(tev.Seed)))
+
+		ci := &crawlInformation[K, N]{
+			todo:    []crawlJob[K, N]{},
+			cpls:    map[string]int{},
+			waiting: map[string]N{},
+			success: map[string]N{},
+			failed:  map[string]N{},
+			errors:  map[string]error{},
+		}
+
+		for _, node := range tev.Seed {
+			// exclude self from closest nodes
+			if key.Equal(node.Key(), c.self.Key()) {
+				continue
+			}
+
+			for j := 0; j < c.cfg.MaxCPL; j++ {
+				target, err := c.cplFn(node.Key(), j)
+				if err != nil {
+					// return nil, fmt.Errorf("generate cpl: %w", err)
+					// TODO: log
+					continue
+				}
+
+				job := crawlJob[K, N]{
+					node:   node,
+					target: target.Key(),
+				}
+
+				ci.cpls[job.mapKey()] = j
+				ci.todo = append(ci.todo, job)
+			}
+		}
+
+		c.info = ci
+
+	case *EventCrawlNodeResponse[K, N]:
+		span.SetAttributes(attribute.Int("closer_nodes", len(tev.CloserNodes)))
+
+		if c.info == nil {
+			return &StateCrawlIdle{}
+		}
+
+		job := crawlJob[K, N]{
+			node:   tev.NodeID,
+			target: tev.Target,
+		}
+
+		mapKey := job.mapKey()
+		if _, found := c.info.waiting[mapKey]; !found {
+			break
+		}
+
+		delete(c.info.waiting, mapKey)
+		c.info.success[mapKey] = tev.NodeID
+
+		for _, node := range tev.CloserNodes {
+			for i := 0; i < c.cfg.MaxCPL; i++ {
+				target, err := c.cplFn(node.Key(), i)
+				if err != nil {
+					// TODO: log
+					continue
+				}
+
+				newJob := crawlJob[K, N]{
+					node:   node,
```

Comment:
What prevents us from crawling a node multiple times for the same target? Nodes A and B could both return node C in their list of nodes closer to target T.

Reply:
Ah, I see that's what `cpls` map is for. Can you add comments on the fields?

Reply:
I think you're right 🤔 Let me write a test for it and assert exactly that 👍